### PR TITLE
Scenario select improvements: tabbed by game or difficulty, campaign mode

### PR DESCRIPTION
**Pull request by AaronVanGeffen**
This PR aims to improve the Scenario Select screen by optionally allowing the user to group scenarios by game source instead of difficulty, as well as "locking" scenarios until enough scenarios have been completed.

As this information is not present in the SV6 files, it has been hardcoded in a new file, `scenario_sources.h`.

**This PR is a work in progress. Please do not merge it yet.**

# Issues

1. Configuring the tab mode and campaign mode through the Options screen is currently not possible. I'm sure I missed something in `options.c` --- I've been staring at it for a long time and would appreciate a hand.
2. Hardcoded scenarios currently look at the scenario name in English. This breaks if a scenario name has been translated, e.g. when playing the game in Dutch. As a workaround, we could keep the English scenario name in memory, but that doesn't seem ideal. Suggestions?
3. If a scenario is not yet unlocked, it is currently listed in a boring grey colour. I'm sure I've seen a proper disabled widget style somewhere in the game, but I can't find it now. What color or style *should* I use, and how?

I may not be able to work on this for the next few days, but I really would appreciate some feedback on what I have so far. PRs are also welcome.

# Screenshots

As it has visible changes, I am including two screenshots. Note that when the options are disabled, the scenario select screen will just look like the regular RCT2 menu instead, including window dimensions.

Without unlocking:

With unlocking ("campaign mode"):

### CODE REVIEW

File: data/language/english_uk.txt [hunk 1]
```diff
@@ -3945,6 +3945,19 @@ STR_5603    :Guest has left ride
 STR_5604    :Guest has bought item
 STR_5605    :Guest has used facility
 STR_5606    :Guest has died
+STR_5607    :RCT1 Classic
+STR_5608    :Corkscrew Follies
```
Comment (IntelOrca) [inline data/language/english_uk.txt]: Change to Added Attractions as that is the correct name for UK. Other languages like US can use CF if they want to.

File: data/language/english_uk.txt [hunk 2]
```diff
@@ -3945,6 +3945,19 @@ STR_5603    :Guest has left ride
...
+STR_5609    :Loopy Landscapes
+STR_5610    :RCT2 Vanilla
```
Comment (IntelOrca) [inline data/language/english_uk.txt]: I don't think saying "Vanilla" is necessary.

File: data/language/english_uk.txt [hunk 3]
```diff
@@ -3945,6 +3945,19 @@ STR_5603    :Guest has left ride
 STR_5604    :Guest has bought item
 STR_5605    :Guest has used facility
 STR_5606    :Guest has died
+STR_5607    :RCT1 Classic
```
Comment (IntelOrca) [inline data/language/english_uk.txt]: I don't think saying "Classic" is necessary.

File: src/localisation/string_ids.h [hunk 4]
```diff
@@ -2229,6 +2229,20 @@ enum {
 	STR_NOTIFICATION_GUEST_USED_FACILITY = 5605,
 	STR_NOTIFICATION_GUEST_DIED = 5606,
 
+	STR_SCENARIO_CATEGORY_RCT1_CLASSIC = 5607,
```
Comment (IntelOrca) [inline src/localisation/string_ids.h]: Same as above, just call it RCT1.

File: src/localisation/string_ids.h [hunk 5]
```diff
@@ -2229,6 +2229,20 @@ enum {
...
+	STR_SCENARIO_CATEGORY_RCT1_CF = 5608,
+	STR_SCENARIO_CATEGORY_RCT1_LL = 5609,
+	STR_SCENARIO_CATEGORY_RCT2_VANILLA = 5610,
```
Comment (IntelOrca) [inline src/localisation/string_ids.h]: and here.

File: src/scenario.h [hunk 6]
```diff
@@ -72,6 +72,17 @@ typedef struct {
 	uint32 scenario_count;		// 0x0C
 } rct_scenario_scores_header;
 
+typedef enum {
+	SCENARIO_SOURCE_RCT1_CLASSIC,
```
Comment (IntelOrca) [inline src/scenario.h]: here...

File: src/scenario.h [hunk 7]
```diff
@@ -72,6 +72,17 @@ typedef struct {
...
+	SCENARIO_SOURCE_RCT1_CORKSCREW,
+	SCENARIO_SOURCE_RCT1_LOOPY,
+	SCENARIO_SOURCE_RCT2_VANILLA,
```
Comment (IntelOrca) [inline src/scenario.h]: here...

File: src/localisation/string_ids.h [hunk 8]
```diff
@@ -2229,6 +2229,20 @@ enum {
 	STR_NOTIFICATION_GUEST_USED_FACILITY = 5605,
 	STR_NOTIFICATION_GUEST_DIED = 5606,
 
+	STR_SCENARIO_CATEGORY_RCT1_CLASSIC = 5607,
+	STR_SCENARIO_CATEGORY_RCT1_CF = 5608,
```
Comment (IntelOrca) [inline src/localisation/string_ids.h]: Use AA instead of CF please.

File: src/scenario.h [hunk 9]
```diff
@@ -72,6 +72,17 @@ typedef struct {
 	uint32 scenario_count;		// 0x0C
 } rct_scenario_scores_header;
 
+typedef enum {
+	SCENARIO_SOURCE_RCT1_CLASSIC,
+	SCENARIO_SOURCE_RCT1_CORKSCREW,
```
Comment (IntelOrca) [inline src/scenario.h]: I think you should either use the abbreviation AA or the full name: Added Attractions... same with the other packs.

File: src/scenario.h [hunk 10]
```diff
@@ -89,6 +100,8 @@ typedef struct {
 	sint32 flags;				// 0x0268
 	uint32 company_value;		// 0x026C
 	char completed_by[64];		// 0x0270
+	scenario_source source_game;// new in OpenRCT2
```
Comment (IntelOrca) [inline src/scenario.h]: We typically don't define occurrences of enums, the enum itself isn't specifying a type size and you might not always want it to use the default. uint8 would be better.

File: src/scenario_list.c [hunk 11]
```diff
@@ -130,6 +184,13 @@ static void scenario_list_add(const char *path)
 		safe_strncpy(scenario->name, s6Info.name, 64);
 		safe_strncpy(scenario->details, s6Info.details, 256);
 	}
+
+	// Normalize the name to make the scenario as recognisable as possible.
+	normalize_scenario_name(scenario);
```
Comment (IntelOrca) [inline src/scenario_list.c]: UK English please for consistency... ```normalise```

**Comment by IntelOrca** [comment]
I also want to add separators in those RCT2 tabs to group the scenarios by difficulty, I will see if I can do that at some point.

**Comment by Gymnasiast** [comment]
All original scenarios use English names: Dutch is at most included via a Scenario text .DAT file. So that shouldn't be an issue. And we use our own translation system anyway.

**Comment by AaronVanGeffen** [comment]
I'm aware of that, @Gymnasiast. The thing is, when OpenRCT2 loads the scenario ([scenario.c, line 92 onwards](https://github.com/OpenRCT2/OpenRCT2/blob/develop/src/scenario.c#L92)) it sets the localised version as its name, so the English version is not preserved or even loaded...

So without any changes, right now, all RCT2 scenarios show up in "Other parks" when playing in Dutch, because they cannot be matched directly.

File: src/scenario.h (hunk 10 above)
Comment (AaronVanGeffen) [inline src/scenario.h]: Yes, I realise it's pretty much a direct copy of the file's header. I didn't notice any problems loading or saving though, so I thought it was fine. Hmm, how else would you have me do it? The scenario list is sorted shortly after loading -- I'd rather not have to keep two arrays sorted consistently...

File: data/language/english_uk.txt (hunk 1 above)
Comment (AaronVanGeffen) [inline data/language/english_uk.txt]: Ah, didn't realise that! I'll change it.

**Comment by IntelOrca** [comment]
@AaronVanGeffen not yet, I will just need to have a play around with it to fix those. I'll do that after you have done my first suggested changes.

**Comment by AaronVanGeffen** [comment]
@IntelOrca Fair enough. I've quickly addressed them in 4be96bdb6e23be60ff16d8280d339f8b3d22c643; hope you find them satisfactory.

**Comment by IntelOrca** [comment]
@AaronVanGeffen :+1:

**Comment by Gymnasiast** [comment]
@AaronVanGeffen : I think it's better to just match on file name then.

**Comment by AaronVanGeffen** [comment]
@Gymnasiast Hmm, I'd really rather not compare by file name -- it would make the system much more inflexible, I think.

If we can't sort it out string-wise, I'd rather store the string-ids for the RCT2 scenarios in the `rct_scenario_basic` struct, and keep a separate array to act as a map to categorise them by these ids.

File: src/scenario.h [hunk 12]
```diff
@@ -89,6 +100,8 @@ typedef struct {
 	sint32 flags;				// 0x0268
 	uint32 company_value;		// 0x026C
 	char completed_by[64];		// 0x0270
+	uint8 source_game;			// new in OpenRCT2
+	sint16 source_index;		// new in OpenRCT2
```
Comment (IntelOrca) [inline src/scenario.h]: If you do this, no one's scores.dat will be read properly any more.

Comment (AaronVanGeffen) [inline src/scenario.h]: Ah, I didn't realise I'd remove backwards compatibility. I did test saving of scores -- those work just fine, but I can see why it breaks.

Do you have any suggestions to do it differently? As I said before, I'd rather not have to keep this data in another array and have it stay sorted consistently with the scenario list so as to not impact performance...

**Comment by IntelOrca** [comment]
@AaronVanGeffen ok, I will take it over from here. I want to make it search multiple locations for scenarios as well.

**Comment by Gymnasiast** [comment]
@IntelOrca : Ok, I promised to add entries to english_uk.txt for the RCT1 scenarios. As far as I'm concerned, that still stands, unless you prefer to add them yourself.

**Comment by IntelOrca** [comment]
@Gymnasiast er, sure you can add them... titles are just a copy and paste job. Descriptions aren't though.

**Comment by Gymnasiast** [comment]
Yes, I remember the SC4/SC4 stuff.
I'll do the names and descriptions then, that way I can also prepare the Dutch translation.

**Comment by IntelOrca** [comment]
Work has now moved to a branch on this repo: **new-scenarioselect**

**Comment by IntelOrca** [comment]
image

**Comment by AaronVanGeffen** [comment]
Looks great, Ted! It's a really nice addition for the RCT2 tabs. It's also nice to see the scenario origin popping up when grouping by difficulty -- big cheers. :+1: 

As for the RCT1 tabs, I think the scenarios for RCT1 are classified quite arbitrarily. I'm not sure if you're aware of this, but for me they're either all beginner (without expansions), challenging (added attractions) or expert (loopy landscapes).

As for the 'other parks' tab, I have two "other parks" subcategories in that tab, so something's not quite right there yet, I guess.

**Comment by Gymnasiast** [comment]
*As for the RCT1 tabs, I think the scenarios for RCT1 are classified quite arbitrarily. I'm not sure if you're aware of this, but for me they're either all beginner (without expansions), challenging (added attractions) or expert (loopy landscapes).*

Crappage did that in an attempt to make sorting more RCT1-like.

**Comment by AaronVanGeffen** [comment]
> Crappage did that in an attempt to make sorting more RCT1-like.

I thought it might be something like that, yes. Unfortunately, we cannot expect the data to be set as such for all conversions, otherwise it'd have made it much easier to implement. It does, however, mean the actual difficulty label shouldn't be used as such for these conversions.

**Comment by Gymnasiast** [comment]
Exactly. Also, when we are able to read the SC*.SC4 files directly, we'll need a hardcoded difficulty anyway. IMO, it would be better to hardcode it now and ignore it for RCT1/AA/LL conversions.

**Comment by IntelOrca** [comment]
> One last thing: I noticed you have changed the code so the "unlocking of scenarios" option has no effect when grouping by difficulty. Any particular reason? There is no obvious indication that the option will have no effect. Personally, I think it should just apply regardless of the mode.

Because the scenarios are sorted by name, and you were locking all but the top 5. This makes no sense and would be affected by new scenarios (suddenly one is no longer unlocked).

**Comment by AaronVanGeffen** [comment]
I'm not too bothered by that, but fair enough. Can we grey out the option when the relevant mode is not selected, then?

**Comment by IntelOrca** [comment]
@AaronVanGeffen yea ok, I have also fixed the categorisation on other parks. You will see the heading "other parks" still but only once. This is because you can have "beginner parks" and "challenging parks" etc. in there as well.

**Comment by AaronVanGeffen** [comment]
Cheers, these headers look great. Good idea to hardcode some difficulty levels for them, for now, too.

One more thing regarding unlocking of scenarios, while you're looking at the code: if you (un)check the box in the options screen with the scenario list open, the window isn't invalidated or closed. I think it can be properly redrawn if invalidated in this case, yes?

**Comment by IntelOrca** [comment]
I've just made it close the window for now. The checkbox now gets disabled when not applicable.

**Comment by AaronVanGeffen** [comment]
@IntelOrca Alright, looks great!  :+1:

I'm not sure if you're looking for bug reports at this point, but I consistently get a malloc-related error (uninitialised pointer being `free`d) after doing the following:

1. Open a scenario
2. Win scenario through cheats
3. Enter name in window
4. Quit to title screen
5. Click new game

**Comment by AaronVanGeffen** [comment]
@IntelOrca Love what you've done in https://github.com/OpenRCT2/OpenRCT2/commit/a53ec386025059c83e6306beffeaa8f49fbceb04!

Found one more bug: "Mythological - Cradle of Civilisation" is currently listed under "Other Parks", not "Time Twister". Strange, as it appears to be in the correct array in `scenario_sources.c`.

**Comment by IntelOrca** [comment]
@AaronVanGeffen Unfortunately, that particular scenario is inconsistent with all the others and does not have title or details. It relies on STEX for all langauges.

**Comment by janisozaur** [comment]
Using 7ef67c2 when I open up the game (vanilla from steam) and click on new scenario, there is no tab selected, just an empty window (it has tabs I can switch to).

zaznaczenie_080

**Comment by IntelOrca** [comment]
@janisozaur fixed :+1:

File: src/scenario_sources.h [hunk 13]
```diff
@@ -0,0 +1,170 @@
...
+	"Geoffrey Gardens",
+
+	 // RCT: Loopy Landscapes
+	"Iceberg Islands",
+	"Volcania",
+	"Arid Heights",
+	"Razor Rocks",
+	"Crater Lake",
+	"Vertigo Views",
+	"Big Pier 2",
```
Comment (Gymnasiast) [inline src/scenario_sources.h]: Should be Paradise Pier 2, with Big Pier 2 defined as alias.

Comment (IntelOrca) [inline src/scenario_sources.h]: This is not the latest version, this file doesn't even exist any more.

**Comment by IntelOrca** [comment]
Moved to #2628.

**Pull request closed without merging**